packages/server-wallet/src/chain-service/event-tracker.ts

```diff
@@ -0,0 +1,105 @@
+import {Address} from '@statechannels/nitro-protocol/src/contract/types';
+import {Logger} from 'pino';
+
+import {
+  HoldingUpdatedArg,
+  AssetOutcomeUpdatedArg,
+  ChannelFinalizedArg,
+  ChallengeRegisteredArg,
+  ChainEventSubscriberInterface,
+} from './types';
+
+/**
+ * This class is a wrapper of ChainEventSubscriberInterface.
+ * Purpose - 2 functions can trigger AssetHolder events to be played to ChainEventSubscriberInterface
+ *    1. chainservice.getInitialHoldings
+ *    2. chainservice.addAssetHolderObservable
+ * Some events can be played twice by both functions, order not predictable,
+ * this class ensures each event is played once and in the right order
+ */
+export class EventTracker {
+  private logger: Logger;
+  private assetHolderMap: Map<Address, {blockNumber: number; logIndex: number}>;
+  private managedSubscriber: ChainEventSubscriberInterface;
+
+  private isNewEvent(assetHolderAddress: Address, blockNumber: number, logIndex: number): boolean {
+    this.logger.debug(
+      `return.isNewEvent: ${assetHolderAddress}, ${blockNumber}, ${logIndex}, ${JSON.stringify(
+        this.assetHolderMap.get(assetHolderAddress) || {}
+      )}`
+    );
+    if (!this.assetHolderMap.has(assetHolderAddress)) {
+      this.assetHolderMap.set(assetHolderAddress, {blockNumber: blockNumber, logIndex: logIndex});
```

Comment:
Nit: it might be nice to factor out this setter logic. One way is to create a small helper function. Another way is to not use early returns and use a mutable variable to determine whether a new value needs to be set.

Reply:
will do